### ISSUE TITLE
"nonce" value not returned

### ISSUE THREAD

**Issue by ajinkyas** [FusionAuth/fusionauth-issues#109]
### Description 
I am requesting on `http://localhost:9011/oauth2/authorize?client_id=75e33455-f1c5-4e29-8863-53ec28364839&redirect_uri=http%3A%2F%2Flocalhost%3A5000%2Fsignin-oidc&response_type=id_token&scope=openid%20profile&response_mode=form_post&nonce=636904122635787810.YzA1ZTZkY2QtNDk5Yy00ZTY0LTlhYzQtMGVhMTc1MjM3NmRmNThhNTk2OGMtZmNiNi00ODExLThjOWMtZGZjYjIxODk5ZjFh&state=CfDJ8NIq6JMx7khEv5e0kR0710V-50-jesO_078ONYha6t2iW9CTel0A_3CGWkKAqkjOJ_o2bCyVdeYCFf13WBVN1Kq_7YXhD4dhfvYzSJqp61pP0gpDvdGubH_M2V7_t8KYF2o5XxwlnX7kfJJxEK_Y4fS_PBT1c-j4OsX4G73BpSb3ZCwP5kLvDNoM-yes2mxEV-zRpAsza4JwZiOWIV5BRuiBVWnGVtYJQW0zdWYeIkV95q9WQWn0opp78sgdR-c4tJgayrMnz5UgzNNCTTs0Q7tWnL-bSqSOLz3K8oGVFu2VhEfMUyi-Ii1-REOvDmImaXuXCX9gMx2AiLRFu2jdrU0&x-client-SKU=ID_NET&x-client-ver=2.1.4.0`

It get redirected to `http://localhost:9011/oauth2/complete-registration?client_id=75e33455-f1c5-4e29-8863-53ec28364839&metaData.device.type=UNKNOWN&redirect_uri=http%3A%2F%2Flocalhost%3A5000%2Fsignin-oidc&response_type=id_token&scope=openid+profile&state=CfDJ8NIq6JMx7khEv5e0kR0710V-50-jesO_078ONYha6t2iW9CTel0A_3CGWkKAqkjOJ_o2bCyVdeYCFf13WBVN1Kq_7YXhD4dhfvYzSJqp61pP0gpDvdGubH_M2V7_t8KYF2o5XxwlnX7kfJJxEK_Y4fS_PBT1c-j4OsX4G73BpSb3ZCwP5kLvDNoM-yes2mxEV-zRpAsza4JwZiOWIV5BRuiBVWnGVtYJQW0zdWYeIkV95q9WQWn0opp78sgdR-c4tJgayrMnz5UgzNNCTTs0Q7tWnL-bSqSOLz3K8oGVFu2VhEfMUyi-Ii1-REOvDmImaXuXCX9gMx2AiLRFu2jdrU0`

Finally it gets redirected to `http://localhost:5000/signin-oidc?code=5thUF9-cgV3InLemn8SpGT5X6iOeo1VoTJLpTeCijB8&state=CfDJ8NIq6JMx7khEv5e0kR0710V-50-jesO_078ONYha6t2iW9CTel0A_3CGWkKAqkjOJ_o2bCyVdeYCFf13WBVN1Kq_7YXhD4dhfvYzSJqp61pP0gpDvdGubH_M2V7_t8KYF2o5XxwlnX7kfJJxEK_Y4fS_PBT1c-j4OsX4G73BpSb3ZCwP5kLvDNoM-yes2mxEV-zRpAsza4JwZiOWIV5BRuiBVWnGVtYJQW0zdWYeIkV95q9WQWn0opp78sgdR-c4tJgayrMnz5UgzNNCTTs0Q7tWnL-bSqSOLz3K8oGVFu2VhEfMUyi-Ii1-REOvDmImaXuXCX9gMx2AiLRFu2jdrU0&userState=Authenticated`
With **id_token**: `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Ijc1ZTMzNDU1LWYxYzUtNGUyOS04ODYzLTUzZWMyODM2NDgzOSJ9.eyJhdWQiOiI3NWUzMzQ1NS1mMWM1LTRlMjktODg2My01M2VjMjgzNjQ4MzkiLCJleHAiOjE1NTQ4MTkwNjMsImlhdCI6MTU1NDgxNTQ2MywiaXNzIjoiYWNtZS5jb20iLCJzdWIiOiJlNjI4NWU1Ni03ZjFkLTQ1YTgtYjFhNS1lOGNiYzc0OTkwZDMiLCJhdXRoZW50aWNhdGlvblR5cGUiOiJQQVNTV09SRCIsImVtYWlsIjoiYWppbmt5YUB0ZWNobm9naXNlLmNvbSIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJhdF9oYXNoIjoiczJVNGt6ZmdGYlo4N0hrY01iVzdPUSIsImNfaGFzaCI6IlVnSlN0RDZZcl94UHZ6MXZjTWt6X0EiLCJhcHBsaWNhdGlvbklkIjoiNzVlMzM0NTUtZjFjNS00ZTI5LTg4NjMtNTNlYzI4MzY0ODM5Iiwicm9sZXMiOltdfQ.DsHUb7LUkZNqhzqfn_vIBsfQIXLgLunAPDbUBjQABIM`
Where **key** (HMAC secret) is: `lOB8KJK2gKn2c6ZFJIKpVqLM-gHa6WmFvSJfHWUKscs`

You can see that `nonce` is **NOT** present in **id_token**

```js
{
  "aud": "75e33455-f1c5-4e29-8863-53ec28364839",
  "exp": 1554819063,
  "iat": 1554815463,
  "iss": "acme.com",
  "sub": "e6285e56-7f1d-45a8-b1a5-e8cbc74990d3",
  "authenticationType": "PASSWORD",
  "email": "ajinkya@technogise.com",
  "email_verified": true,
  "at_hash": "s2U4kzfgFbZ87HkcMbW7OQ",
  "c_hash": "UgJStD6Yr_xPvz1vcMkz_A",
  "applicationId": "75e33455-f1c5-4e29-8863-53ec28364839",
  "roles": []
}
```

### Expected behavior
`nonce` should be present in **id_token** payload data.

### Platform
I am using `Microsoft.AspNetCore.Authentication.OpenIdConnect`
 - Device: Desktop
 - OS: macOS
 - Browser + version: chrome, safari

### Additional context
`Self service registration` is turned ON and I need it to be ON.

**Comment by ajinkyas** [comment]
Thanks for the reply. I can see nonce in id_token after disabling self service registration. 

But I ran into another issue now. It is redirecting back to this URL, which is different than earlier:

> http://localhost:5000/signin-oidc#scope=openid+profile&state=CfDJ8NIq6JMx7khEv5e0kR0710X-wNYZqFYu2MDXjO3mDGBe4pnGUptGEHETF7uX-I-UZnCCn1NxC1CnKoeRX_sbvRvu26u1zS0waDq9diyIMzeXrnl6-_bXPkXa0tHsUp6of23Axj8tEqpnYqHPoo6IzSjo0H5TU5eptxfcem_nq81gIHoC9Zw7bPf7tO3oOgKJtNqxZTChSl0UEKCIpGo5vmUhvUKY83KY2BHq5dPDzbVBv5iOw8jjc0rhX0EJDr152Aw4UVBuDlc4goj80ckVseN7pvH1Q_s3nr4sJkuAxuYjKNEPfmrg1k_58mCbQ-ASg1PhpR6rPg9VZ913dPl6S0E&userState=Authenticated&id_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Ijc1ZTMzNDU1LWYxYzUtNGUyOS04ODYzLTUzZWMyODM2NDgzOSJ9.eyJhdWQiOiI3NWUzMzQ1NS1mMWM1LTRlMjktODg2My01M2VjMjgzNjQ4MzkiLCJleHAiOjE1NTQ4Nzg3NjEsImlhdCI6MTU1NDg3NTE2MSwiaXNzIjoiYWNtZS5jb20iLCJzdWIiOiJlNjI4NWU1Ni03ZjFkLTQ1YTgtYjFhNS1lOGNiYzc0OTkwZDMiLCJhdXRoZW50aWNhdGlvblR5cGUiOiJQQVNTV09SRCIsImVtYWlsIjoiYWppbmt5YUB0ZWNobm9naXNlLmNvbSIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJhcHBsaWNhdGlvbklkIjoiNzVlMzM0NTUtZjFjNS00ZTI5LTg4NjMtNTNlYzI4MzY0ODM5Iiwibm9uY2UiOiI2MzY5MDQ3MTk2MTg4OTE3OTAuTTJRNVlXVmlNekl0TTJJMk5TMDBOall3TFdJNFkySXRNVGcxTkdNMU9UY3dZMk5qTW1JMk5XTmhaR010WlROaVlpMDBPVEU0TFdJMk9XTXRNRGN3WVdFME9UWXdaV05rIiwicm9sZXMiOltdfQ.3D5tb__7j527vjjJUOKePH6qSnDkWfulUlzrMS7S8LA

Look at the `#` at `http://localhost:5000/signin-oidc#` in above URL. Why is the URL changing after disabling self service registration? You can compare it with the one mentioned in the issue's description above. I started getting below error after disabling self service registration, irrespective of whether I keep nonce or not. Is this an expected behavior?
> System.Exception: Correlation failed.
   at Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler`1.HandleRequestAsync()
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

**Comment by robotdan** [comment]
I think what you're observing is the issue described in this issue https://github.com/FusionAuth/fusionauth-issues/issues/102. 

In your authorization request you provided `response_type=id_token` which indicates you're asking for the Implicit Grant. The Implicit Grant will return the `id_token` on the redirect URL using a URL fragment (`#`). 

So when you disabled self-service registration, you bypassed the bug, and you are now getting the correct redirect URL. The URL you received previously which used a `?` instead of a `#` was incorrect according to the grant you requested based upon your `response_type`.

**Comment by robotdan** [comment]
Fixed in 1.6.0
